# Read CSV and JSON

I wrote this so I can reopen it later and remember how a file becomes a DataFrame.

**How I revise:** run every code cell from the top. The CSV is large, so I use `shape`, `head`, and `tail` instead of scrolling the whole table.

## The picture I keep

| I call | File | What comes back |
| --- | --- | --- |
| `pd.read_csv` | `data/titanic.csv` | The full passenger table |
| `pd.read_json` | `data/sample.json` | The same kind of table, only 6 rows |
| `df.head()` | | First 5 rows |
| `df.tail()` | | Last 5 rows |
| `df.shape` | | `(rows, columns)` |

Both calls return a DataFrame. The file format changes. The object I get does not.

In [18]:
import pandas as pd

## Where the files are

This notebook lives in `notebooks/`. The files live in `data/`, one folder up.

`../data/titanic.csv` means: leave this folder, then open `data/titanic.csv`.

If I see `FileNotFoundError`, the kernel was started from a different folder.

## CSV — the full table

`read_csv` reads comma-separated text. The first line of the file is the column names.

I store that table in `df`. I do not reuse `df` for the JSON file, or I lose this table.

In [19]:
# ../data/ means: leave notebooks/, then open the file in data/.
df = pd.read_csv("../data/titanic.csv")
print("shape (rows, columns):", df.shape)

shape (rows, columns): (891, 12)


## First look: head, tail, shape

I do not read all 891 rows to check the load.

- `head()` is the first 5 rows. I use it to see the column names and a few passengers.
- `tail()` is the last 5 rows. On this file, that is the end of the full voyage.
- `shape` is `(rows, columns)`. This is the number I remember. The CSV has many more rows than the 6-row JSON.

In [20]:
# First 5 rows. Enough to see the column names and a few passengers.
print(df.head())

   PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S  


In [21]:
# Last 5 rows. These are the end of the full file, not the six-row sample.
print(df.tail())

     PassengerId  Survived  Pclass                                      Name  \
886          887         0       2                     Montvila, Rev. Juozas   
887          888         1       1              Graham, Miss. Margaret Edith   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   
889          890         1       1                     Behr, Mr. Karl Howell   
890          891         0       3                       Dooley, Mr. Patrick   

        Sex   Age  SibSp  Parch      Ticket   Fare Cabin Embarked  
886    male  27.0      0      0      211536  13.00   NaN        S  
887  female  19.0      0      0      112053  30.00   B42        S  
888  female   NaN      1      2  W./C. 6607  23.45   NaN        S  
889    male  26.0      0      0      111369  30.00  C148        C  
890    male  32.0      0      0      370376   7.75   NaN        Q  


In [22]:
# (rows, columns). The CSV is the full voyage, so the row count is much larger than 6.
print(df.shape)

(891, 12)


## JSON — six passengers

`sample.json` is a list of objects, one object per passenger. `read_json` turns that list into a DataFrame.

The sample is a short cut of the same voyage: 6 rows, and fewer columns than the CSV. One age is missing. Pandas shows a missing number as `NaN`, not as zero.

I store it in `df2` on purpose. If I write `df = pd.read_json(...)`, I replace the CSV table.

In [23]:
# df2, not df, so the full CSV table stays in df.
df2 = pd.read_json("../data/sample.json")
print("json shape:", df2.shape)
print("csv shape:", df.shape)
print(df2)
print("missing age:", df2.loc[5, "Age"])

json shape: (6, 8)
csv shape: (891, 12)
   PassengerId  Survived  Pclass                          Name     Sex   Age  \
0            1         0       3       Braund, Mr. Owen Harris    male  22.0   
1            2         1       1    Cumings, Mrs. John Bradley  female  38.0   
2            3         1       3        Heikkinen, Miss. Laina  female  26.0   
3            4         1       1  Futrelle, Mrs. Jacques Heath  female  35.0   
4            5         0       3      Allen, Mr. William Henry    male  35.0   
5            6         0       3              Moran, Mr. James    male   NaN   

      Fare Embarked  
0   7.2500        S  
1  71.2833        C  
2   7.9250        S  
3  53.1000        S  
4   8.0500        S  
5   8.4583        Q  
missing age: nan


## Check myself before I close this

I should be able to say these without looking up:

1. `read_csv` and `read_json` both give me a DataFrame.
2. The path starts with `../` because this notebook is not in the same folder as the data.
3. `head` is the start, `tail` is the end, `shape` is `(rows, columns)`.
4. `NaN` means missing. In the JSON, the last passenger has no age.
5. `df` is the CSV. `df2` is the JSON. One load must not erase the other.

**If I only remember two lines:**

```python
df = pd.read_csv("../data/titanic.csv")
df2 = pd.read_json("../data/sample.json")
```